### ISSUE TITLE
xgboost-sklearn custom scoring cutoff

### ISSUE THREAD

**Issue by geoHeil** [dmlc/xgboost#1890]
Similar to xgboost4j I want to use a custom scoring cutoff in xgboost's sklearn wrapper https://github.com/dmlc/xgboost/issues/1763

So far I did not find this functionality. Is this missing?

**Comment by geoHeil** [comment]
@xydrolase do you know if it is possible to set the thresholds for the python API as well?

**Comment by xydrolase** [comment]
Never used it before. What is it for?

**Comment by geoHeil** [comment]
Generally, xgboost performs a 50% cut-off in case of binary classification. Apparently, this can be modified via that parameter in xgboost4j / scala version.

I will asl @CodingCat. He helped me thit #1763

**Comment by xydrolase** [comment]
I followed your exchanges in #1763. So basically the thresholds are used as scaling factors for scaling the predicted probabilities. 

If you use `clf.predict_proba()` to get the probabilities, and then scale the probabilities (columns) accordingly using the same thresholds/coefficients, and identify the argmax in each row, it should achieve the same goal. No?

**Comment by geoHeil** [comment]
I am not entirely sure about that.
Xin Yin <notifications@github.com> schrieb am Mo. 30. Jan. 2017 um 14:53:

**Comment by geoHeil** [comment]
@xydrolase may I add https://github.com/dmlc/xgboost/issues/1749 as well. When directly `predict_proba` I get the following output. But that does not really assemble probabilities. I would have expected a more histogram like plot (with different (sort of gaussian) bars.

**Comment by xydrolase** [comment]
If you look at the link (https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassificationModel.scala#L127) in @CodingCat 's first reply, the probabilities are clearly scaled using the thresholds.

With regard to your quest in #1749, `predict_proba()` returns multiple columns, each corresponding to a class label. If you're performing binary classification, then `clf.predict_proba()[:, 1]` should give you the probability of the positive label.

**Comment by geoHeil** [comment]
Indeed that is what I would have expected. Just saying that I would have expected a more "randomly" distributed probability per class and not sort of a straight line. That's where I am not sure if the result is correct.

**Comment by xydrolase** [comment]
Oh, I see what you meant. Sorry about the earlier confusion. 
I guess it has to do with your dataset. I meant, if you have an awesome that correlates with the class label almost perfectly, then surely the model can perform exceptionally well. 

I used `predict_proba()` before in many tasks, and I didn't really see that kind of behavior.

**Comment by geoHeil** [comment]
Strange. Because metrics e.g. kappa are still nowhere near 1 (0.5-0.6) so I thought it should be more distributed.

**Comment by xydrolase** [comment]
What is kappa? Sorry I'm not familiar with the metric.

**Comment by geoHeil** [comment]
It is https://en.wikipedia.org/wiki/Cohen%27s_kappa which goes from -1 to 1, greater is better and seemed to work fairly well for an highly imbalanced classification task.

**Comment by xydrolase** [comment]
Thanks for the link. It's an interesting metric. I never used it before. Since I'm not familiar with the properties of the kappa metric, I'm not sure if 0.5~0.6 is indicative or not about what you've observed on the probability distribution.

**Comment by geoHeil** [comment]
Thank you for the good explanation. Will close the issue assuming it is related to my dataset as you said.
